operator/pkg/ciliumenvoyconfig/manager.go

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2019-2020 Authors of Cilium
+
+package ciliumenvoyconfig
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/watchers/subscriber"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var _ subscriber.Service = &manager{}
+
+const (
+	ciliumEnvoyLBPrefix = "cilium-envoy-lb"
+)
+
+type svcEvent string
+
+type manager struct {
+	envoyConfigManager *envoyConfigManager
+
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	serviceStore cache.Store
+	ports        []string
+	algorithm    string
+}
+
+// New returns a new manager for CiliumEnvoyConfig
+func New(ctx context.Context, indexer cache.Store, ports []string, algorithm string) (*manager, error) {
+	manager := &manager{
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		serviceStore: indexer,
+		maxRetries:   10,
+		ports:        ports,
+		algorithm:    algorithm,
+	}
+
+	envoyConfigManager, err := newEnvoyConfigManager(manager.maxRetries)
+	if err != nil {
+		return nil, err
+	}
+	manager.envoyConfigManager = envoyConfigManager
+
+	go manager.run(ctx)
+	return manager, nil
+}
+
+func (m *manager) OnAddService(service *slim_corev1.Service) error {
+	var (
+		svcName   = service.Name
+		scopedLog = log.WithField(logfields.ServiceName, svcName)
+	)
+	key, err := cache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		return err
+	}
+
+	if !IsLBProtocolAnnotationEnabled(service) && !hasAnyPort(service, m.ports) {
+		return nil
```

Comment:
What if the service had l7lb disabled while the controller was being upgraded? If so, we'll leak the config here. You can't shortcut this case :-)

Reply:
can you elaborate a little bit more here ? This is for service added event, also the annotation will have higher priority i.e. if the service annotation is disabled, we will not do anything even the port requirement is satisfied.